### PR TITLE
Adding the ability to update the repo owner and name via patch with t…

### PR DESCRIPTION
**Pull request by josmo**
…he owner and name props

@bradrydzewski It looks like I've got everything working from being able to patch. I'll add a cli PR that's missing the model till that gets merged that way if you have any comments I can address it right.

The cli has the following warning when not unsafe of "validate updating the full repo name is unsafe - build status in repos will reflect old name"
since that seems to be the only side affect.  I also have the cli repair the hooks after it renames it.

for https://github.com/drone/drone/issues/2167

### CODE REVIEW

File: model/repo.go [hunk 1]
```diff
@@ -67,4 +67,6 @@ type RepoPatch struct {
 	AllowDeploy  *bool   `json:"allow_deploy,omitempty"`
 	AllowTag     *bool   `json:"allow_tag,omitempty"`
 	BuildCounter *int    `json:"build_counter,omitempty"`
+	Name         *string `json:"name",omitempty`
```
Comment (mrueg) [inline model/repo.go]: not sure if the quoting is right here: shouldn't it be "name,omitempty"?

Comment (josmo) [inline model/repo.go]: You're right it should be "name,omitempty" I'm not sure why it worked while testing but I'll fix and push

File: server/repo.go [hunk 2]
```diff
@@ -137,6 +138,19 @@ func PatchRepo(c *gin.Context) {
...
+		from, err := remote.Repo(user, *in.Owner, *in.Name)
+		if err != nil {
+			c.AbortWithError(http.StatusInternalServerError, err)
+		}
+		repo.Name = from.Name
+		repo.Owner = from.Owner
+		repo.FullName = from.FullName
+		repo.Avatar = from.Avatar
+		repo.Link = from.Link
+		repo.Clone = from.Clone
```
Comment (bradrydzewski) [inline server/repo.go]: maybe also update `private` and `visibility`

Comment (josmo) [inline server/repo.go]: Ok I'll totally add that.  Right now I have the CLI run a repair after the update which seems to work seamlessly from the user perspective. At least with what I've tried it works.

**Comment by josmo** [comment]
@bradrydzewski Let me know what you lean towards. Right now in the CLI I'm doing a update and then repair as part of one command, but I guess both of those could take place on the server as well and part of the repair endpoint?

**Comment by bradrydzewski** [comment]
I think I am in favor for including this in the repair endpoint directly. When you change the name of a repository it is sort of in a broken state, and the repair function fixes the state. So maybe something like this:

```
drone repo repair foo/bar --renamed=foo/baz
```

We should also fetch the user permissions for the new repository, to ensure they have admin access to the repository with the provided name.

**Comment by josmo** [comment]
I think it can make sense to use the repair endpoint.  On the CLI I tend to like the``` drone repo mv --unsafe currentproject/currentrepo newproject/newrepo ``` since it's a bit more intuitive from a user. Sound reasonable?

**Comment by bradrydzewski** [comment]
Yes I like the idea of adding a new move command and endpoint

```
POST /api/repos/{owner}/{repo}/move?to={owner}/{repo}
```

The endpoint should probably handle updating the webhook, since some people might use the API directly and not the CLI

**Comment by josmo** [comment]
@bradrydzewski Let me know if this seems reasonable.  Looks like it works on my side.

File: server/repo.go [hunk 3]
```diff
@@ -235,3 +220,53 @@ func RepairRepo(c *gin.Context) {
...
+		c.AbortWithError(http.StatusInternalServerError, err)
+	}
+	repo.Name = from.Name
+	repo.Owner = from.Owner
+	repo.FullName = from.FullName
+	repo.Avatar = from.Avatar
+	repo.Link = from.Link
+	repo.Clone = from.Clone
+	repo.IsPrivate = from.IsPrivate
+	repo.Visibility = from.Visibility
```
Comment (bradrydzewski) [inline server/repo.go]: this needs to be calculated using the following logic

```
	if repo.Visibility == "" {
		repo.Visibility = model.VisibilityPublic
		if repo.IsPrivate {
			repo.Visibility = model.VisibilityPrivate
		}
	}
```

Comment (bradrydzewski) [inline server/repo.go]: maybe we add a `repo.ResetVisibility()` function?

File: server/repo.go [hunk 4]
```diff
@@ -235,3 +220,53 @@ func RepairRepo(c *gin.Context) {
...
+	if errStore != nil {
+		c.AbortWithError(http.StatusInternalServerError, errStore)
+		return
+	}
+
+	RepairRepo(c)
+}
+
+// ParseRepo parses the repository owner and name from a string.
+func ParseRepo(str string) (user, repo string, err error) {
```
Comment (bradrydzewski) [inline server/repo.go]: this might be nice to add to `model/repo.go` so that we can re-use in other places?

File: server/repo.go [hunk 5]
```diff
@@ -235,3 +220,53 @@ func RepairRepo(c *gin.Context) {
...
+
+func MoveRepo(c *gin.Context) {
+	remote := remote.FromContext(c)
+	repo := session.Repo(c)
+	user := session.User(c)
+
+	to, exists := c.GetQuery("to")
+	if !exists {
+		err := fmt.Errorf("Missing required to query value")
+		c.AbortWithError(http.StatusInternalServerError, err)
```
Comment (bradrydzewski) [inline server/repo.go]: I think we need to always `return` after an `AbortWithError`.  If you see any examples in the code where we are not, they are probably bugs :)

File: server/repo.go (hunk 3 above)
Comment (josmo) [inline server/repo.go]: Sounds good I'll add since I'll need to use it when I create a PR to have patch behave the same way

File: server/repo.go [hunk 6]
```diff
@@ -235,3 +220,53 @@ func RepairRepo(c *gin.Context) {
...
+	}
+	repo.Name = from.Name
+	repo.Owner = from.Owner
+	repo.FullName = from.FullName
+	repo.Avatar = from.Avatar
+	repo.Link = from.Link
+	repo.Clone = from.Clone
+	repo.IsPrivate = from.IsPrivate
+	repo.Visibility = from.Visibility
+
```
Comment (bradrydzewski) [inline server/repo.go]: we should fetch the user permissions for the repository and verify they have admin access, and aren't trying to rename their repository to something they don't have access to.

File: server/repo.go (hunk 3 above)
Comment (bradrydzewski) [inline server/repo.go]: note we should probably only re-calculate visibility if the `Private` value changed

```
if repo.IsPrivate != from.IsPrivate {
  repo.ResetVisibility()
}
```

The visibility value is set by the user, and lets you mark a private repository as publicly visible and vice versa. I think it makes sense to change the visibility automatically when a repository private value is changed.

File: server/repo.go (hunk 5 above)
Comment (josmo) [inline server/repo.go]: oops, I think it's just me being stupid :), fixed in the next commit

File: server/repo.go (hunk 4 above)
Comment (josmo) [inline server/repo.go]: done in the next commit

File: server/repo.go (hunk 3 above)
Comment (josmo) [inline server/repo.go]: done in the next commit

File: server/repo.go (hunk 6 above)
Comment (josmo) [inline server/repo.go]: I think what I have works. But Let me know if there's a better/preferred way

**Comment by josmo** [comment]
@bradrydzewski I think all the comments you add are taken care of :)

**Comment by bradrydzewski** [comment]
thanks, I think we are good to go :)

**Pull request merged**